Switch resample to the new drizzle package API

mcara opened the pull request:
Resolves [JP-3685](https://jira.stsci.edu/browse/JP-3685)
Resolves [JP-3713](https://jira.stsci.edu/browse/JP-3713)

Closes #8638 
Closes #8721 

This PR switches resample step to the new `drizzle` API proposed in https://github.com/spacetelescope/drizzle/pull/134

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [ ] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [ ] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.fits_generator.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

emolter commented on jwst/resample/gwcs_drizzle.py:
```diff
@@ -1,7 +1,6 @@
 
-from drizzle import util
-from drizzle import cdrizzle
+from drizzle import cdrizzle, utils
```
I think our pre-commit is complaining because `utils` is unused and should be removed from the imports

emolter commented on jwst/resample/gwcs_drizzle.py:
```diff
@@ -381,9 +380,13 @@ def dodrizzle(insci, input_wcs, inwht, output_wcs, outsci, outwht, outcon,
     # Compute the mapping between the input and output pixel coordinates
     # for use in drizzle.cdrizzle.tdriz
+
+    # pixmap = utils.calc_pixmap(input_wcs, output_wcs)
```
remove commented line

emolter commented on jwst/resample/gwcs_drizzle.py:
```diff
@@ -381,9 +380,13 @@ def dodrizzle(insci, input_wcs, inwht, output_wcs, outsci, outwht, outcon,
+    # pixmap = utils.calc_pixmap(input_wcs, output_wcs)
     pixmap = resample_utils.calc_gwcs_pixmap(input_wcs, output_wcs, insci.shape)
+
     # inwht[np.isnan(pixmap[:,:,0])] = 0.
```
you didn't modify this, but is there any reason this commented line needs to stay in this file>

emolter commented on jwst/resample/gwcs_drizzle.py:
```diff
@@ -381,9 +380,13 @@ def dodrizzle(insci, input_wcs, inwht, output_wcs, outsci, outwht, outcon,
+
     # inwht[np.isnan(pixmap[:,:,0])] = 0.
 
+    print(f"Pixmap shape: {pixmap.shape}")
```
remove print statement

emolter commented on jwst/resample/resample.py:
```diff
@@ -290,7 +290,7 @@ def resample_group(self, input_models, indices):
                     # must call this explicitly to populate area extension
                     # although the existence of this extension may not be necessary
-                    img.area = img.area 
+                    img.area = img.area
```
what does this line do? does it break anything to remove it?

mcara replied:
I have no idea. It is a strange line but in Python, sometimes things are magical. Seems to be related to SlitModel and also used (set) in the `photom` step (multi-slit). Maybe @melanieclarke knows more.

emolter commented on jwst/resample/resample_utils.py:
```diff
@@ -5,6 +5,7 @@
 import numpy as np
 from astropy import units as u
 import gwcs
```
I think our pre-commit is complaining because gwcs is unused

emolter commented on jwst/resample/resample_utils.py:
```diff
@@ -124,17 +125,11 @@ def shape_from_bounding_box(bounding_box):
-
-    grid = gwcs.wcstools.grid_from_bounding_box(bb)
-    pixmap = np.dstack(reproject(in_wcs, out_wcs)(grid[0], grid[1]))
+    if shape and not np.array_equiv(shape, in_wcs.array_shape):
```
again this isn't your fault as you didn't modify this, but it's best practice to use `is` or `is not` when comparing against `None`; see https://docs.astral.sh/ruff/rules/none-comparison/

mcara replied:
I do not find this harmful but I will update.

emolter commented on jwst/resample/resample_utils.py:
```diff
@@ -124,17 +125,11 @@ def shape_from_bounding_box(bounding_box):
     """
-    if shape:
-        bb = wcs_bbox_from_shape(shape)
-        log.debug("Bounding box from data shape: {}".format(bb))
```
would this debug message still be helpful to keep around?

mcara replied:
I don't see what for.

emolter commented on pyproject.toml:
```diff
@@ -22,7 +22,7 @@ dependencies = [
     "BayesicFitting>=3.0.1",
     "crds>=11.17.14",
-    "drizzle>=1.15.0",
+    "drizzle @ git+https://github.com/mcara/drizzle.git@redesign-api",
```
remember to change to @main before merge

mcara replied:
so changed

emolter reviewed (commented):
Thanks Mihai, looks good overall and my comments are minor.  It needs a changelog entry.  It would also be good to run our regression test suite; do you have permissions and know how to do that?

mcara commented:
Regression tests running here: https://github.com/spacetelescope/RegressionTests/actions/runs/11247877551

nden commented on jwst/resample/resample_utils.py:
```diff
@@ -124,17 +124,11 @@ def shape_from_bounding_box(bounding_box):
-    grid = gwcs.wcstools.grid_from_bounding_box(bb)
-    pixmap = np.dstack(reproject(in_wcs, out_wcs)(grid[0], grid[1]))
+    if shape is not None and not np.array_equiv(shape, in_wcs.array_shape):
+        in_wcs = deepcopy(in_wcs)
```
Is a `deepcopy` necessary here?

mcara replied:
I am not sure. It felt to be a safer way because this method sets `array_shape` property of the WCS object. If there is no drawback in modifying inputs, `deepcopy` can be omitted.

nden replied:
What happens here if `shape is None` ?

braingram commented:
Is the new drizzle API intended to make direct calls to `tdriz` `tblot` unnecessary?

mcara commented:
> Is the new drizzle API intended to make direct calls to `tdriz` `tblot` unnecessary?

In general, new drizzle API is not "intended" to rely on `tdriz` and such but these calls are available to the developers. That is, the Python API was the one that was modified, not the C-based function API.

mcara commented:
> Is the new drizzle API intended to make direct calls to `tdriz` `tblot` unnecessary?

In general, new drizzle API is not "intended" to rely on `tdriz` and such but these calls are available to the developers. That is, the Python API was the one that was modified, not the C-based function API.

The purpose of this PR is to adapt the current code that otherwise would break to the new API when needed. We know that we are going to switch to the new API anyway via other PRs in `jwst` and `stcal` when the resample code is migrated there. So this PR simply switches to the new `drizzle` code with minimal effort in order to understand and OKify (when reasonable) regression test changes due to (minor) bug fixes and the Python class redesign.

CC: @nden

emolter commented:
This looks mostly good to me now, but there are 12 failing tests in the regression test run and only 5 failing tests in last night's nightly run.  Do these changes look to you like they are caused by this PR, and are any of them large?

nden commented:
@mcara I may be confused about what this PR is meant to accomplish.
- How is it using the new API in drizzle? The only change here is calling `calc_pixmap` from drizzle with a modified signature. The rest of the changes are simply reformatting lines (I prefer these not be included) or fixing white spaces.
- Changing the signature of `calc_pixmap` means the input WCS needs to be modified to include the change in shape, which means a deepcopy of the WCS is needed. Deep copies of WCS objects are usually expensive and I would prefer the old signature in order to avoid the copies.
- The new implementation of `calc_pixmap` computes the input pixels in a different way and does not follow the adopted convention (using `grid_from_bounding_box`) . This means the inputs are probably slightly off the center of pixels and this will lead to different results. Is there a reason not to use the convention? If you are concerned that gwcs will become a dependency of drizzle then the function should go in stcal (there might even be one there already).

mcara commented:
After the first commit, regression tests are passing, see https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1787

mcara commented:
After commit https://github.com/spacetelescope/jwst/pull/8866/commits/91805f78087e824fe4eeeb81c01ffaeb81013ee3, regression tests were started here: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1792/

mcara commented:
Regression tests after complete switch to the new `Drizzle` class + performance improvements (commit 0858ed5d8d07c6c2b676e1cc6eb7a25f21a4cc32): https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1798/

emolter commented:
I'll review this in detail later, but I wanted to add a note about the memory usage in outlier detection that we discussed earlier.  Here is the [JIRA ticket](https://jira.stsci.edu/browse/JP-3685) and [GitHub Issue](https://github.com/spacetelescope/jwst/issues/8638), and I'm also putting a flamegraph here of the current memory usage in outlier detection coming from GWCSDrizzle.  The memory usage has been largely unchanged by your PR.  I think the con and wht arrays are coming from the fact that `blank_output` has those populated, and they are each taking up (32 bits * output size) of memory.  They are useless though - I'm seeing that both are fully populated with zeros.

I don't really care whether a change that fixes this is part of this PR or the next one, but I do feel it should be fixed by one of them.

<img width="755" alt="Screenshot 2024-10-17 at 12 17 24 PM" src="https://github.com/user-attachments/assets/e663ef23-4fb4-42d3-ba14-03229743f1ce">

emolter reviewed (commented):
Just a few comments.  The biggest one is to check whether setting the context array to None is really working as intended.

The ResampleStep memory usage is also roughly unchanged by this PR, but I think the refactor in the second PR is much more likely to make improvements

mcara commented on jwst/resample/gwcs_drizzle.py:
```diff
@@ -55,70 +57,61 @@ def __init__(self, product, outwcs=None, wt_scl=None,
+
+        # Since the context array is dynamic, it must be re-assigned
+        # back to the product's `con` attribute.
+        product.con = self.out_ctx
```
in my tests it worked - it simply creates an zeroes array of the same shape as the data array.

emolter replied:
Ok, but that means that the context array is still pre-allocated and takes up memory even though it doesn't do anything.

Also, while this may work as you intended, I would argue that it's unclear, and it would be easy to replace it with something like `model.con = np.zeros_like(model.con)`

mcara replied:
I don't think we need to replace anything if context is `None`, i.e., I'll have `if con is not None then assign`

nden commented:
@mcara Please resolve the conflicts in resample.

mcara commented:
> I'll review this in detail soon, but I wanted to add a note about the memory usage in outlier detection that we discussed earlier. Here is the [JIRA ticket](https://jira.stsci.edu/browse/JP-3685) and [GitHub Issue](https://github.com/spacetelescope/jwst/issues/8638), and I'm also putting a flamegraph here of the current memory usage in outlier detection coming from GWCSDrizzle. The memory usage has been largely unchanged by your PR. I think the con and wht arrays are coming from the fact that `blank_output` has those populated, and they are each taking up (32 bits * output size) of memory. The con array is useless though - I'm seeing that it is fully populated with zeros both before and after drizzling. I don't know if the weights are needed or not here, but they do appear nontrivial so I guess they are.
> 
> I don't really care whether a change that fixes this is part of this PR or the next one, but I do feel it should be fixed by one of them.

This seems to be a property of data models so that part should be fixed in order to have some memory savings here. For outlier detection, there isn't much that can be done in the resample step. In one of your PRs where you switched to the `ModelLibrary` you basically already implemented the main saving: don't load all input image files into the memory. Creation of context array _by the resample_ is turned off in this PR but data models create a full `0s` array anyway.

Main performance improvements in this PR apply to execution time and not to the memory saving. For example, for one of the data sets I run the resample step on, I got a reduction of the running time from 18.5 min to 11.5 min - a reduction of about 40% with this PR.

braingram commented on jwst/resample/gwcs_drizzle.py:
```diff
@@ -55,70 +48,56 @@ def __init__(self, product, outwcs=None, wt_scl=None,
-
-    def add_image(self, insci, inwcs, inwht=None, xmin=0, xmax=0, ymin=0, ymax=0,
-                  expin=1.0, in_units="cps", wt_scl=1.0, iscale=1.0):
+        ctx = product.con
```
The `product.con` access here will create a context array (if one doesn't already exist). Putting it in an `if not disable_ctx` block (and similarly guarding other `product.con` accesses would prevent this). See https://github.com/spacetelescope/jwst/issues/8638#issuecomment-2422454499 for more details.

mcara commented:
Latest regression tests are running here: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1800/

emolter commented:
I just tested the most recent commit, and it does appear that the context array is no longer allocated.  Thanks for updating that.  This does modestly decrease the total memory usage of outlier detection.  I'll put a note on https://github.com/spacetelescope/jwst/issues/8638 about investigating further improvements

mcara commented:
So any approvals?

emolter commented on jwst/resample/gwcs_drizzle.py:
```diff
@@ -55,70 +48,60 @@ def __init__(self, product, outwcs=None, wt_scl=None,
+                if ctx.size == 0:
+                    ctx = None
+                elif ctx.ndim not in [2, 3]:
+                    # TODO: this message seems odd
```
do you have a more descriptive message to suggest?  Comment can be removed if not

emolter commented on pyproject.toml:
```diff
@@ -22,7 +22,7 @@ dependencies = [
     "BayesicFitting>=3.0.1",
     "crds>=11.17.14",
-    "drizzle>=1.15.0",
+    "drizzle>=2.0.0",
```
Did this release already happen?

emolter reviewed (commented):
Looks mostly good to me.  Wondering about the drizzle release schedule, it looks like with the new pin that release needs to be made before we can merge this.

Also I wanted to be sure you were aware of these two JIRA tickets related to resample performance:
https://jira.stsci.edu/browse/JP-3713
https://jira.stsci.edu/browse/JP-3707
Does this PR resolve (fully or partially) either of those, and if so, can you please update those tickets and add something to the top-level summary of this PR?

mcara commented:
dropped the commit that was removing the pin

mcara commented:
> Also I wanted to be sure you were aware of these two JIRA tickets related to resample performance:
> https://jira.stsci.edu/browse/JP-3713
> https://jira.stsci.edu/browse/JP-3707
> Does this PR resolve (fully or partially) either of those, and if so, can you please update those tickets and add something to the top-level summary of this PR?

Thanks for mentioning this. No, this PR does not address those two issues.

However, it can be addressed either via their own PRs or it can be incorporated into this PR. I would prefer separate PRs for clarity.

emolter commented:
Sorry to continue along this line so long, but I have one more question about memory usage.  I'm not going to hold up the PR for this by any means, but I am curious if you think any additional memory usage improvements can be had here or if it's not possible.  And maybe @braingram may have some answers or ideas here too.

See the flamegraph attached to [this JIRA ticket](https://jira.stsci.edu/browse/JP-3685) with filename `memray-flamegraph-outlier_AL875_mihai_ondisk.html`.  Note that for this test dataset, the size of the output drizzled array takes 479 MB of memory as a float32.

Why does `self.blank_output` allocate 479 MB of memory, whereas self.blank_output.copy() allocates double that (958 MB)?  I don't see anywhere that model explicitly gets new extensions.

Why does `self.outwht` allocate 958 MB of memory instead of 479 MB?  It looks like this is 2 allocations each of 479 MB.  Is the reason because product.wht is not populated before this line?  i.e., the same behavior that was causing the `.con` extension to allocate memory unnecessarily?  Is there any way to store this just once?

emolter reviewed (approved):
I'm not an official maintainer, so this needs at least a few more approvals, but it looks good to me

braingram commented:
> Sorry to continue along this line so long, but I have one more question about memory usage. I'm not going to hold up the PR for this by any means, but I am curious if you think any additional memory usage improvements can be had here or if it's not possible. And maybe @braingram may have some answers or ideas here too.

I think there are more memory improvements that are possible (like not keeping `blank_output` to copy and instead creating a new model when it's needed). That might fit into the resample refactor PR.

> See the flamegraph attached to [this JIRA ticket](https://jira.stsci.edu/browse/JP-3685) with filename `memray-flamegraph-outlier_AL875_mihai_ondisk.html`. Note that for this test dataset, the size of the output drizzled array takes 479 MB of memory as a float32.
> 
> Why does `self.blank_output` allocate 479 MB of memory, whereas self.blank_output.copy() allocates double that (958 MB)? I don't see anywhere that model explicitly gets new extensions.

How many groups are in the test association you're using? Since memray is reporting the allocations it's likely that because the copy is called more than once python doesn't reuse the memory for the second call (although the third might reuse the first, etc). Looking at the scope of `output_model` it's created in `resample_group` and becomes `drizzled_model` here:
https://github.com/spacetelescope/jwst/blob/030cd930d0693991df31be29744c88a08827e72f/jwst/outlier_detection/utils.py#L225C13-L225C27
Since this is in a loop where the `drizzled_model` from a previous iteration doesn't lose all references until the line that calls `resample_group` the memory can't be reused. My guess is adding a `del drizzled_model` to the end of that loop would reduce the allocations to match.

> Why does `self.outwht` allocate 958 MB of memory instead of 479 MB? It looks like this is 2 allocations each of 479 MB. Is the reason because product.wht is not populated before this line? i.e., the same behavior that was causing the `.con` extension to allocate memory unnecessarily? Is there any way to store this just once?

Same as above.

mcara commented:
In https://github.com/spacetelescope/jwst/pull/8866/commits/ca1d76514ca673f942d49379f996e5071be63e29 I modified the code to no longer store `iscale` in the data model. For the `resample_variance_arrays` to work, I introduced a dictionary of `iscale` values: `self._iscales`. `resample_variance_arrays` is no longer used anywhere and so maybe if we could remove it from the `ResampleData` class, we can remove as well `self._iscales`.

Also, `GWCSDrizzle` was modified to take `output_wcs` instead of an output model. Now output models are created when needed with empty arrays and only in the end output model arrays are set.

Maybe this will provide some additional memory savings, CC: @emolter

nden commented:
In the future, can `GWCSDrizzle` be removed and `ResampleData` use directly `drizzle.Drizzle`? The only additional info in `GWCSDrizzle` is the output wcs.

mcara commented:
> In the future, can `GWCSDrizzle` be removed and `ResampleData` use directly `drizzle.Drizzle`? The only additional info in `GWCSDrizzle` is the output wcs.

I was thinking about the same. The only thing is `add_image` prints extra debug things that, I would suggest, we should just get rid of:
```python
        log.debug(f"Pixmap shape: {pixmap[:,:,0].shape}")
        log.debug(f"Input Sci shape: {insci.shape}")
        log.debug(f"Output Sci shape: {self.out_img.shape}")

        # Call 'drizzle' to perform image combination
        log.info(f"Drizzling {insci.shape} --> {self.out_img.shape}")
```
Maybe keep the last `info`

mcara commented:
Regression tests running here:

emolter commented:
> Maybe this will provide some additional memory savings

This seems to make little difference for memory.  Now the memory allocation has indeed moved inside of Drizzle, but it's still making two allocations within `_alloc_output_arrays` (I assume this is `self._out_img` and `self._out_wht`) and it's doing so two times, for a total of 4x the output size as float32.  I still think there must be a way to get this down to 2x the output size, i.e., allocate out_img and out_wht once.  I see there are lots of `del` statements on your branch but it doesn't look like they solve this.  I'm testing Brett's suggestion of deleting `drizzled_model` in the `median_with_resampling` loop now.

mcara commented:
> ... it's still making two allocations within `_alloc_output_arrays` (I assume this is `self._out_img` and `self._out_wht`) and it's doing so two times, for a total of 4x the output size as float32.

Why do you say that? Is there an example I could run to debug this? It should not be allocating output image and weight arrays twice.

emolter commented:
It's happening for the reason Brett indicated in his comment.  Adding the lines
```
            if i != ngroups - 1:
                del drizzled_model
```
at the end of the `for` loops [here](https://github.com/spacetelescope/jwst/blob/3842f77fbded258014a10c4979ad9bd0325ff254/jwst/outlier_detection/utils.py#L139) and [here](https://github.com/spacetelescope/jwst/blob/3842f77fbded258014a10c4979ad9bd0325ff254/jwst/outlier_detection/utils.py#L247) solves the problem, and in conjunction with my proposed fix for AL-875, brings the total memory usage of outlier detection in on-disk=True mode for my use-case down from 4 GB on main to 2 GB.

melanieclarke commented:
Looking at the regtest failures for metadata in NIRSpec FS spec3, it looks like the way the intermediate median output model is made is under-determined.  The regtest appears to be picking a different model for the top-level update in this branch than it did previously, but I think the issue is in the [jwst.outlier_detection.utils.median_with_resampling function](https://github.com/spacetelescope/jwst/blob/3842f77fbded258014a10c4979ad9bd0325ff254/jwst/outlier_detection/utils.py#L164).  Probably the `indices_by_group` list need to be sorted, and the first drizzled model needs to be used to update the output model, instead of the last.  

Tagging @emolter - this is related to #8782

emolter commented:
> Probably the indices_by_group list need to be sorted, and the first drizzled model needs to be used to update the output model, instead of the last.

Agree sorting it is a safe option.  But why should it matter which of those models is used to update the metadata?  They are all equally valid inputs to the median.  One reason I hesitate to make that change is that it makes [this memory saving option](https://github.com/spacetelescope/jwst/pull/8866#issuecomment-2427149300) more challenging to implement.  If there are pieces of metadata that depend on which of the input drizzled models you choose to copy, wouldn't it be more correct to either update those in a smart way using, say, ModelBlender or else remove them entirely?

melanieclarke commented:
> Agree sorting it is a safe option. But why should it matter which of those models is used to update the metadata? They are all equally valid inputs to the median. One reason I hesitate to make that change is that it makes [this memory saving option](https://github.com/spacetelescope/jwst/pull/8866#issuecomment-2427149300) more challenging to implement. If there are pieces of metadata that depend on which of the input drizzled models you choose to copy, wouldn't it be more correct to either update those in a smart way using, say, ModelBlender or else remove them entirely?

The issue is that the version on main is taking metadata from the first file, made when the blank_output is created, then writing it to a filename corresponding to the second file, which is a bit unexpected.  This version makes metadata and filename match, but both are from the later file instead of the earlier one.  I think that's also a little unexpected, but as you say, not terribly important since this is an intermediate file used for diagnostic purposes only.

I think you can still delete the drizzled_model as suggested, even if you use the first model, though.  You just need to make a blank ImageModel at the top of the function for the median model, update the metadata and wcs for it when i=0, and update the data and err arrays at the end.  I can show you what I mean offline if that's not clear enough.

emolter commented:
> The issue is that the version on main is taking metadata from the first file, made when the blank_output is created, then writing it to a filename corresponding to the second file, which is a bit unexpected. This version makes metadata and filename match, but both are from the later file instead of the earlier one. I think that's also a little unexpected, but as you say, not terribly important since this is an intermediate file used for diagnostic purposes only.

> I think you can still delete the drizzled_model as suggested, even if you use the first model, though. You just need to make a blank ImageModel at the top of the function for the median model, update the metadata and wcs for it when i=0, and update the data and err arrays at the end. I can show you what I mean offline if that's not clear enough.

I think it is clear what you mean, and thanks for figuring out the problem.  It is not clear what you are suggesting to do about it though.  This is a bug that is separate from this PR, so should I change the behavior in a separate PR?  Or should we just rely on merging this PR branch to fix it, as it sounds like the behavior on this PR branch is already acceptable?

melanieclarke commented:
> I think it is clear what you mean, and thanks for figuring out the problem. It is not clear what you are suggesting to do about it though. This is a bug that is separate from this PR, so should I change the behavior in a separate PR? Or should we just rely on merging this PR branch to fix it, as it sounds like the behavior on this PR branch is already acceptable?

Either way! If you're deleting the drizzled models in the loop in this PR, I think we might as well also change the behavior to take metadata reliably from the first input.  If not, we can approve the regtest changes as is in this PR and deal with the median output separately.

emolter commented:
I'm not doing anything in this PR, this is all up to Mihai

mcara commented:
@emolter @melanieclarke Would [this change](https://github.com/spacetelescope/jwst/pull/8866/commits/e0d4a8d14a79d74efe3a4107b7675a53563256bb) save memory and fix the issue of meta being taken from the last image for the median model?

melanieclarke commented:
> @emolter @melanieclarke Would [this change](https://github.com/spacetelescope/jwst/pull/8866/commits/e0d4a8d14a79d74efe3a4107b7675a53563256bb) save memory and fix the issue of meta being taken from the last image for the median model?

Thanks, Mihai, that's exactly what I had in mind.  With this change, you'll also need to update the filename that the regtest looks for, here:
[test_nirspec_fs_spec3](https://github.com/spacetelescope/jwst/blob/232d2647ed58d3af8c1b5317de7c81cbb3dea902/jwst/regtest/test_nirspec_fs_spec3.py#L39) 
and update the associated truth files.  It will now make `jw01309022001_04102_00001*median.fits` instead of `jw01309022001_04102_00002*median.fits`.

mcara commented:
sigh... I missed that. will do. Thanks!

emolter commented:
Yeah that looks fine to me for memory saving

mcara commented:
Finally, I think regression tests now pass: https://github.com/spacetelescope/RegressionTests/actions/runs/11449076833/job/31854019443

There are two failing tests (starfinder alternatives/tweakreg source finding) that fail but I can't see how they are related to this PR.

I believe this is ready to merge after getting required approvals. CC: @nden @tapastro

mcara commented:
... well, after `drizzle` PR is merged and a release is made. Not to forget, https://github.com/spacetelescope/jwst/blob/90ba645e6c57d809222fff75909cf71be3e7971c/pyproject.toml#L25 must be modified to require the new `drizzle` version.

mcara commented:
Also, `romancal` regression tests do not conflict with the new `drizzle` - see https://github.com/spacetelescope/RegressionTests/actions/runs/11410576103/job/31764795945

I spoke to @schlafly and he did not object to making a release of the `drizzle` package.

mcara commented:
One more thing: before merging, would it be OK to remove `resample_variance_arrays()` that's no longer used - see https://github.com/spacetelescope/jwst/blob/90ba645e6c57d809222fff75909cf71be3e7971c/jwst/resample/resample.py#L704-L749 and `self._iscales` that's needed only to support `resample_variance_arrays()` (for backward compatibility - I highly doubt that anyone would be calling this function at all by itself)?

emolter commented:
If it's no longer used, yes, I think you can and should remove it

melanieclarke commented:
> One more thing: before merging, would it be OK to remove `resample_variance_arrays()` that's no longer used

I agree with Ned - that function should be removed since it's no longer used.  It was just a helper function to compute the variance arrays while resampling.  Since the relevant calculations have been moved inside the loop over models instead of doing them after the loop has completed, there's no reason to keep the old function.

mcara commented:
Regression tests after the last two commits (pin drizzle + remove `_iscales` and `resample_variance_arrays()`): https://github.com/spacetelescope/RegressionTests/actions/runs/11494300632
